cpp/src/arrow/dataset/file_parquet_test.cc

```diff
@@ -367,6 +369,69 @@ TEST_F(TestParquetFileFormat, MultithreadedScan) {
   ASSERT_EQ(batches.size(), kNumRowGroups);
 }
 
+class AsyncBufferReader : public ::arrow::io::BufferReader {
+ public:
+  explicit AsyncBufferReader(std::shared_ptr<Buffer> buffer,
+                             const ::arrow::io::IOContext& ctx)
+      : ::arrow::io::BufferReader(std::move(buffer)), ctx_(ctx) {}
+  /// EXPERIMENTAL: Read data asynchronously.
+  Future<std::shared_ptr<Buffer>> ReadAsync(const ::arrow::io::IOContext& ctx,
+                                            int64_t position, int64_t nbytes) override {
+    auto self = checked_pointer_cast<AsyncBufferReader>(shared_from_this());
+    return DeferNotOk(ctx.executor()->Submit(
+        [self, position, nbytes]() { return self->ReadAt(position, nbytes); }));
+  }
+
+  const ::arrow::io::IOContext& io_context() const override { return ctx_; }
+
+ private:
+  ::arrow::io::IOContext ctx_;
+};
+
+TEST_F(TestParquetFileFormat, SingleThreadExecutor) {
+  ::arrow::io::IOContext default_io_context;
+  auto pool_executor =
+      dynamic_cast<::arrow::internal::ThreadPool*>(default_io_context.executor());
+  if (pool_executor == nullptr) {
+    GTEST_SKIP();
+  }
+  auto origin_capacity = pool_executor->GetCapacity();
+  ASSERT_OK(pool_executor->SetCapacity(/*threads=*/1));
+
+  // Reset capacity for pool_executor
+  struct PoolResetGuard {
+    PoolResetGuard(::arrow::internal::ThreadPool* pool, int origin_capacity)
+        : pool(pool), origin_capacity(origin_capacity) {}
+    ~PoolResetGuard() {
+      Status s = pool->SetCapacity(origin_capacity);
+      if (!s.ok()) {
+        std::cerr << "Failed to reset pool capacity: " << s.ToString() << std::endl;
+      }
+    }
+
+    ::arrow::internal::ThreadPool* pool;
+    int origin_capacity;
+  };
+  PoolResetGuard guard(pool_executor, origin_capacity);
```

Comment:
I think you can use `SetIOThreadPoolCapacity` and `GetIOThreadPoolCapacity` here:

```suggestion
  // Reset capacity for pool_executor
  struct PoolResetGuard {
    int original_capacity = io::GetIOThreadPoolCapacity();
    ~PoolResetGuard() {
      DCHECK_OK(io::SetIOThreadPoolCapacity(original_capacity));
    }
  } guard;
```

Since this directly mutates the IO thread pool (pointed to by the default IO context), you shouldn't need `AsyncBufferReader::ReadAsync`

I agree that it's odd `BufferReader::ReadAsync` and `MemoryMappedFile::ReadAsync` (the only overrides I see) ignore the io context argument. @lidavidm do you think they should transfer to the other executor if it's different?

Reply:
> I agree that it's odd BufferReader::ReadAsync and MemoryMappedFile::ReadAsync (the only overrides I see) ignore the io context argument

I think `BufferReader::ReadAsync` is so lightweight that we can ignore it?

Reply:
Given both access memory/"memory" I think that's quite intentional